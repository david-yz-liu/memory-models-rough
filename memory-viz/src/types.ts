interface DrawnObject {
    isClass: boolean;
    name?: string;
    type?: string;
    x?: number;
    y?: number;
    id: number | string;
    value: any;
    stack_frame: boolean;
    show_indexes?: boolean;
    style?: Style;
}

interface AttributeStyle {
    [propName: string]: string | number;
}

interface Style {
    text_id?: AttributeStyle;
    text_type?: AttributeStyle;
    text_value?: AttributeStyle;
    box_id?: AttributeStyle;
    box_type?: AttributeStyle;
    box_container?: AttributeStyle;
}
