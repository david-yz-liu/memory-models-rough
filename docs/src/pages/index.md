---
title: Home
sidebar_position: 1
---

# Memory model diagrams

This package generates memory model diagrams for Python code in the style of CSC110/111/148 at the University of Toronto.
This uses the [Rough.js](https://roughjs.com/) Javascript library to emulate the look of hand-drawn diagrams.

**Note**: this project is currently experimental, and may undergo significant changes before stabilizing.

## Installation

You can install the `memory-models-rough` package from GitHub (it is currently not on npm):

```console
$ npm install git+https://github.com/david-yz-liu/memory-models-rough.git
```

## Example

Here's an example of using this package in a Javascript file executed by NodeJS.

```javascript
// simple_demo.js
const { draw } = require("memory_models_rough");

const objects = [
    {
        isClass: true,
        name: "__main__",
        id: null,
        value: { lst1: 82, lst2: 84, p: 99, d: 10, t: 11 },
        stack_frame: true,
    },
    {
        isClass: false,
        name: "str",
        id: 19,
        value: "David is cool!",
        style: ["highlight"],
    },
    { isClass: false, name: "int", id: 13, value: 7 },
];

const m = draw(objects, true, { width: 1300 });

m.save("simple_demo.svg");
```

Running `$ node simple_demo.js` produces the following file:

![Diagram generated for simple_demo.js file.](images/simple_demo.svg)
