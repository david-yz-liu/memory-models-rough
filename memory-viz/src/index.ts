import { MemoryModel } from "./memory_model";
import { config } from "./config";
import {
    drawAutomated,
    drawAutomatedOtherItems,
    drawAutomatedStackFrames,
    getSize,
} from "./automate";
import { draw } from "./user_functions";

export default {
    MemoryModel,
    config,
    drawAutomated,
    drawAutomatedOtherItems,
    drawAutomatedStackFrames,
    draw,
    getSize,
};
