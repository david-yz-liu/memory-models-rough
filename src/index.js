const {MemoryModel, config} = require("./memory_model.js");
const {drawAutomated, drawAutomatedOtherItems, drawAutomatedStackFrames, getSize} = require("./automate.js");
const {draw} = require("./user_functions.js");

export default {MemoryModel, config, drawAutomated, drawAutomatedOtherItems, drawAutomatedStackFrames, draw, getSize};
