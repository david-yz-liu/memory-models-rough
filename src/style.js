const merge = require("deepmerge");
const {config} = require("./config");

// Built-in style for drawing text on canvas (if no style is provided by the user).
const default_text_style = {'fill': config.text_color, 'text-anchor': 'middle',
    'font-family': 'Consolas, Courier', 'font-size': config.font_size};

// Default style attributes that apply universally to any type of data.
const common_style = {
    "text_id" :{"fill": config.id_color,'text-anchor': 'middle',
        'font-family': 'Consolas, Courier', 'font-size': config.font_size},
    "text_type" : {"fill": config.value_color, 'text-anchor': 'middle',
        'font-family': 'Consolas, Courier', 'font-size': config.font_size},
    "text_value": {'text-anchor': 'middle', 'font-family': 'Consolas, Courier',
        'font-size': config.font_size},
    "box_container":{},
    "box_id": {},
    "box_type": {}
};

const category_specific_styles = {
    "collection": {
        text_value: {"fill": config.id_color}
    },
    "primitive": {
        text_value: {"fill": config.value_color}
    },
    "class": {
        text_value: {"fill": config.value_color, "text-anchor" : "begin"}
    },
    "stackframe": {
        text_value: {"fill": config.text_color,  "text-anchor" : "begin"}
    }
}

const immutable = ["int", "str", "tuple", "None", "bool", "float", "date"]
const collections = ["list", "set", "tuple", "dict"]

const primitives = ["int", "str", "None", "bool", "float", "date"]

/**
 * Populates a user-passed style object --to the extent needed-- with default data (to adhere to the interface of the
 * style object). Needed to avoid errors of the type "TypeError: Cannot set properties of undefined (setting 'x')", as
 * well as many more.
 * @param {Object} object : the object that represents a Python object the user wants drawn. The style object
 *                          corresponding to 'object' will be extracted be doing object.style.
 * @returns {Object}
 */
function populateStyleObject(object) {

    let style_so_far = common_style;


    let object_type;


    if (primitives.includes(object.name)) {
        object_type = "primitive"
    } else if (collections.includes(object.name)) {
        object_type = "collection"
    } else if (object.stack_frame) {
        object_type = "stackframe"
    } else {
        object_type = "class"
    }

    // We then add properties specific to the different type categories.
    // Note that, the later will take precedence over styleSoFar.
    style_so_far = merge(style_so_far, category_specific_styles[object_type]);


    // Finally, we complement the current style with any user-supplied properties.
    // Note that, the later will take precedence over styleSoFar.
    style_so_far = merge(style_so_far, object.style || {})

    return style_so_far;
}


// Constants employed to establish presets for styles.
const HIGHLIGHT_TEXT = {'font-weight': "bolder", 'font-size': "22px"};
const FADE_TEXT = {/*'font-weight': "normal",*/ "fill-opacity": .4};
const HIGHLIGHT_BOX_LINES = {"roughness": 0.2, "strokeWidth" : 4}
const HIGHLIGHT_BOX = {"roughness": 0.2, "strokeWidth" : 4, "fill":"yellow", "fillStyle" : "solid"};
const FADE_BOX_LINES = {'roughness': 2.0, "strokeWidth": 0.5}
const FADE_BOX = {'roughness': 2.0, "strokeWidth": 0.5, "fill":"rgb(247, 247, 247)", "fillStyle" : "solid"};
const HIDE_BOX = {"fill" : "white", "fillStyle" : "solid"};

const presets = {
    "highlight": {
        "text_value" : HIGHLIGHT_TEXT,
        "text_id": HIGHLIGHT_TEXT,
        "text_type": HIGHLIGHT_TEXT,
        "box_id": HIGHLIGHT_BOX_LINES,
        "box_type": HIGHLIGHT_BOX_LINES,
        "box_container": HIGHLIGHT_BOX
    },
    "highlight_id": {
        "text_id": HIGHLIGHT_TEXT,
        "box_id": HIGHLIGHT_BOX
    },
    "highlight_type": {
        "text_type": HIGHLIGHT_TEXT,
        "box_type": HIGHLIGHT_BOX
    },
    "fade" : {
        "text_value" : FADE_TEXT,
        "text_id": FADE_TEXT,
        "text_type": FADE_TEXT,
        "box_id": FADE_BOX_LINES,
        "box_type": FADE_BOX_LINES,
        "box_container": FADE_BOX
    },
    "fade_id" : {
        "text_id": FADE_TEXT,
        "box_id": FADE_BOX_LINES,
    },
    "fade_type" : {
        "text_type": FADE_TEXT,
        "box_type": FADE_BOX_LINES,
    },
    "hide": {
        "box_container": HIDE_BOX,
        "box_id": HIDE_BOX,
        "box_type": HIDE_BOX
    },
    "hide_id" : {
        "box_id": HIDE_BOX
    },
    "hide_type": {
        "box_type": HIDE_BOX
    },
    "hide_container": {
        "box_container": HIDE_BOX
    }
}

export {
    populateStyleObject,
    immutable,
    collections,
    primitives,
    presets,
    default_text_style}