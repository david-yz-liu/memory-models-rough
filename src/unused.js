/**
 * Makes the memory model more visually appealing by a "vertical centering" of the objects' boxes.
 *
 * The idea is that the first object in a row (the "rowBreaker") could be much taller than the remaining ones in that
 * row, so instead of putting all y-coordinates of the top-left corner to be the same across all these objects,
 * the non-rowBreaker boxes are brought down a bit so that they look vertically centered. Specifically, they are
 * brought down until the margin from the top of the rowBreaker and the margin from the bottom of the rowBreaker
 * are equal.
 *
 * NOTE: Once again, there is no canvas here or actual visual displacements, but merely mutation of the passed list
 * of objects, changing the "y" property of some objects. Of course, this does translate to visual changes once
 * you put this list of objects on canvas (e.g. using the 'MemoryModel.drawAll' method).
 *
 * @param {[object]} objs the list of objects which we want to vertically beautify
 * @returns {[object]} a mutates list of objects, with altered y-coordinates for the objects.
 */
function vert_beautify(objs) {
    // Initially, the height against which we compare is the height of the first object, the tallest of all.
    let height = objs[0].height

    for (const obj of objs) {
        // Ensuring every object has the 'rowBreaker' property. All "first" objects in row have them from the
        // 'drawAutomatedOtherItems' functions, but the rest do not.
        obj.rowBreaker = obj.rowBreaker|false;

        // we are changing row, so we now need to compare against the first object of that row.
        // All displacement happen relative to the "rowBreaker" object, so this object does not move at all.
        if (obj.rowBreaker) {
            height = obj.height;

        } else {
            // Calculations (to make margins from the top and the bottom of the "rowBreaker" box equal)
            const diff = height - obj.height;
            const displacement = diff/2;
            obj.y = obj.y + displacement
        }
    }

    return objs;
}


/**
 *
 * @param {Array} objects
 */
function warnForBlankSpaces(objects) {
    for (let i=0; i < objects.length; i++) {
        const obj = objects[i];

        if ((obj.name === "BLANK") && (!obj.hasOwnProperty("height") || !obj.hasOwnProperty("width"))) {
            objects.splice(i);
            console.log('%c Oh my heavens! ', 'background: #222; color: #bada55');
        }
    }
}