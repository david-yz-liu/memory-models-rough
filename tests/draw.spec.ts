import { MemoryModel } from "../src/memory_model";
const { draw } = require("../dist/memory_models_rough");

describe("draw function", () => {
    let objects: Array<object>;
    beforeEach(() => {
        objects = [
            {
                isClass: true,
                name: "__main__",
                id: null,
                value: { lst1: 82, lst2: 84, p: 99, d: 10, t: 11 },
                stack_frame: true,
            },
            {
                isClass: false,
                name: "str",
                id: 19,
                value: "David is cool!",
                style: ["highlight"],
            },
            { isClass: false, name: "int", id: 13, value: 7 },
        ];
    });

    it("should produce consistent svg when provided seed", () => {
        const m: MemoryModel = draw(objects, true, {
            width: 1300,
            seed: 12345,
        });
        const svg = m.serializeSVG();
        expect(svg).toMatchSnapshot();
    });
});
