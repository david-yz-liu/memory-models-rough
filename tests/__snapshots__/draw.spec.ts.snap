// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`draw function should produce consistent svg when provided seed 1`] = `"<svg width="1300" height="575" xmlns="http://www.w3.org/2000/svg"><g><path d="M23.917262064293027 24.157105488702655 C76.22765391450375 24.75776646597311, 127.45038592889905 26.521927158664912, 222.07396535761654 24.381777508184314 M25.22362225409597 25.469827466644347 C101.08440828286112 25.048337728809564, 179.59881121683867 26.439913750719278, 222.2794330334291 24.511956655420363 M222.3301306789082 24.634793578576517 C224.05348183822807 144.46320941640022, 222.60876481321927 264.58611205202584, 222.18424611596677 474.9097838327125 M223.40175848590926 25.46480132583862 C223.22626858997106 180.44371499744182, 222.9826775629485 335.6344560233484, 223.43236907946323 475.2168707690574 M223.83588658459485 476.08132497780025 C168.846978119947 476.86122407080603, 113.45754360854625 475.0490183389373, 24.056711124256253 473.50267897360027 M222.39905177708715 474.6283317739144 C174.51738550215958 477.8186496617831, 127.29635517541318 476.59090767527, 24.334270439110696 475.56836631242186 M25.068692949442973 475.61772256167717 C26.832202358664823 357.60491966672384, 27.36860749856539 239.13449660839086, 25.857870498417128 24.484317093179772 M25.327713088196433 474.5473643299796 C22.679788707843922 342.0183882892476, 23.279992381386723 209.6133689213003, 25.455403787352658 24.628507300103422" stroke="#000" stroke-width="1" fill="none"/></g><g><path d="M146.91726206429303 99.15710548870265 C163.457930556871 99.4688816714473, 178.91093921363353 101.23304236413911, 205.07396535761654 99.38177750818431 M148.22362225409597 100.46982746664435 C169.6097253277898 99.39177608871833, 193.649445306696 100.78335211062804, 205.2794330334291 99.51195665542036 M204.61389331705868 99.24430773966014 C208.00123269716278 112.83524710033089, 205.0117960111238 127.03554429486394, 204.31202613003552 149.8133229445666 M206.8313265061006 100.96177598182112 C205.952838587109 117.12844209931791, 205.44879527622834 133.73342551570386, 206.89466654416174 150.44875323120505 M206.83588658459485 151.08132497780025 C190.5881071029231 151.29606103045865, 173.9398015744984 149.4838552985899, 147.05671112425625 148.50267897360027 M205.39905177708715 149.6283317739144 C191.23452811203896 151.42153838021682, 177.73064039517195 150.19379639370368, 147.3342704391107 150.56836631242186 M148.14214079268277 151.27820358984172 C148.56834565615281 137.0871600760147, 149.67828564858064 121.94919998198748, 149.77512239106 98.93293885700405 M148.67811032291502 149.0633974308148 C146.80644848523661 134.39918356761336, 148.04840164957568 119.99146327283233, 148.94233041163534 99.23130004946142" stroke="rgb(0, 0, 0)" stroke-width="1" fill="none"/></g><text text-anchor="start" font-family="Consolas, Courier" font-size="20" fill="rgb(0, 0, 0)" x="50" y="130">lst1</text><text fill="rgb(150, 100, 28)" text-anchor="middle" font-family="Consolas, Courier" font-size="20" x="177" y="130">id82</text><g><path d="M146.91726206429303 174.15710548870265 C163.457930556871 174.46888167144732, 178.91093921363353 176.23304236413912, 205.07396535761654 174.3817775081843 M148.22362225409597 175.46982746664435 C169.6097253277898 174.39177608871833, 193.649445306696 175.78335211062804, 205.2794330334291 174.51195665542036 M204.61389331705868 174.24430773966014 C208.00123269716278 187.8352471003309, 205.0117960111238 202.03554429486394, 204.31202613003552 224.8133229445666 M206.8313265061006 175.96177598182112 C205.952838587109 192.1284420993179, 205.44879527622834 208.73342551570386, 206.89466654416174 225.44875323120505 M206.83588658459485 226.08132497780025 C190.5881071029231 226.29606103045865, 173.9398015744984 224.4838552985899, 147.05671112425625 223.50267897360027 M205.39905177708715 224.6283317739144 C191.23452811203896 226.42153838021682, 177.73064039517195 225.19379639370368, 147.3342704391107 225.56836631242186 M148.14214079268277 226.27820358984172 C148.56834565615281 212.0871600760147, 149.67828564858064 196.94919998198748, 149.77512239106 173.93293885700405 M148.67811032291502 224.0633974308148 C146.80644848523661 209.39918356761336, 148.04840164957568 194.99146327283233, 148.94233041163534 174.23130004946142" stroke="rgb(0, 0, 0)" stroke-width="1" fill="none"/></g><text text-anchor="start" font-family="Consolas, Courier" font-size="20" fill="rgb(0, 0, 0)" x="50" y="205">lst2</text><text fill="rgb(150, 100, 28)" text-anchor="middle" font-family="Consolas, Courier" font-size="20" x="177" y="205">id84</text><g><path d="M146.91726206429303 249.15710548870265 C163.457930556871 249.46888167144732, 178.91093921363353 251.23304236413912, 205.07396535761654 249.3817775081843 M148.22362225409597 250.46982746664435 C169.6097253277898 249.39177608871833, 193.649445306696 250.78335211062804, 205.2794330334291 249.51195665542036 M204.61389331705868 249.24430773966014 C208.00123269716278 262.8352471003309, 205.0117960111238 277.03554429486394, 204.31202613003552 299.8133229445666 M206.8313265061006 250.96177598182112 C205.952838587109 267.1284420993179, 205.44879527622834 283.73342551570386, 206.89466654416174 300.44875323120505 M206.83588658459485 301.08132497780025 C190.5881071029231 301.29606103045865, 173.9398015744984 299.4838552985899, 147.05671112425625 298.50267897360027 M205.39905177708715 299.6283317739144 C191.23452811203896 301.4215383802168, 177.73064039517195 300.1937963937037, 147.3342704391107 300.56836631242186 M148.14214079268277 301.2782035898417 C148.56834565615281 287.0871600760147, 149.67828564858064 271.9491999819875, 149.77512239106 248.93293885700405 M148.67811032291502 299.0633974308148 C146.80644848523661 284.39918356761336, 148.04840164957568 269.99146327283233, 148.94233041163534 249.23130004946142" stroke="rgb(0, 0, 0)" stroke-width="1" fill="none"/></g><text text-anchor="start" font-family="Consolas, Courier" font-size="20" fill="rgb(0, 0, 0)" x="50" y="280">p</text><text fill="rgb(150, 100, 28)" text-anchor="middle" font-family="Consolas, Courier" font-size="20" x="177" y="280">id99</text><g><path d="M146.91726206429303 324.15710548870265 C163.457930556871 324.4688816714473, 178.91093921363353 326.2330423641391, 205.07396535761654 324.3817775081843 M148.22362225409597 325.46982746664435 C169.6097253277898 324.3917760887183, 193.649445306696 325.783352110628, 205.2794330334291 324.51195665542036 M204.61389331705868 324.24430773966014 C208.00123269716278 337.8352471003309, 205.0117960111238 352.03554429486394, 204.31202613003552 374.8133229445666 M206.8313265061006 325.9617759818211 C205.952838587109 342.1284420993179, 205.44879527622834 358.73342551570386, 206.89466654416174 375.44875323120505 M206.83588658459485 376.08132497780025 C190.5881071029231 376.29606103045865, 173.9398015744984 374.4838552985899, 147.05671112425625 373.50267897360027 M205.39905177708715 374.6283317739144 C191.23452811203896 376.4215383802168, 177.73064039517195 375.1937963937037, 147.3342704391107 375.56836631242186 M148.14214079268277 376.2782035898417 C148.56834565615281 362.0871600760147, 149.67828564858064 346.9491999819875, 149.77512239106 323.93293885700405 M148.67811032291502 374.0633974308148 C146.80644848523661 359.39918356761336, 148.04840164957568 344.99146327283233, 148.94233041163534 324.2313000494614" stroke="rgb(0, 0, 0)" stroke-width="1" fill="none"/></g><text text-anchor="start" font-family="Consolas, Courier" font-size="20" fill="rgb(0, 0, 0)" x="50" y="355">d</text><text fill="rgb(150, 100, 28)" text-anchor="middle" font-family="Consolas, Courier" font-size="20" x="177" y="355">id10</text><g><path d="M146.91726206429303 399.15710548870265 C163.457930556871 399.4688816714473, 178.91093921363353 401.2330423641391, 205.07396535761654 399.3817775081843 M148.22362225409597 400.46982746664435 C169.6097253277898 399.3917760887183, 193.649445306696 400.783352110628, 205.2794330334291 399.51195665542036 M204.61389331705868 399.24430773966014 C208.00123269716278 412.8352471003309, 205.0117960111238 427.03554429486394, 204.31202613003552 449.8133229445666 M206.8313265061006 400.9617759818211 C205.952838587109 417.1284420993179, 205.44879527622834 433.73342551570386, 206.89466654416174 450.44875323120505 M206.83588658459485 451.08132497780025 C190.5881071029231 451.29606103045865, 173.9398015744984 449.4838552985899, 147.05671112425625 448.50267897360027 M205.39905177708715 449.6283317739144 C191.23452811203896 451.4215383802168, 177.73064039517195 450.1937963937037, 147.3342704391107 450.56836631242186 M148.14214079268277 451.2782035898417 C148.56834565615281 437.0871600760147, 149.67828564858064 421.9491999819875, 149.77512239106 398.93293885700405 M148.67811032291502 449.0633974308148 C146.80644848523661 434.39918356761336, 148.04840164957568 419.99146327283233, 148.94233041163534 399.2313000494614" stroke="rgb(0, 0, 0)" stroke-width="1" fill="none"/></g><text text-anchor="start" font-family="Consolas, Courier" font-size="20" fill="rgb(0, 0, 0)" x="50" y="430">t</text><text fill="rgb(150, 100, 28)" text-anchor="middle" font-family="Consolas, Courier" font-size="20" x="177" y="430">id11</text><g><path d="M23.917262064293027 24.157105488702655 C52.72183570805937 24.567927886713296, 80.43874951601029 26.3320885794051, 130.07396535761654 24.381777508184314 M25.22362225409597 25.469827466644347 C65.28675948381425 24.616882936749608, 108.00351361874492 26.008458958659322, 130.2794330334291 24.511956655420363 M129.61389331705868 24.244307739660144 C133.00123269716278 37.83524710033089, 130.0117960111238 52.03554429486394, 129.31202613003552 74.81332294456661 M131.8313265061006 25.96177598182112 C130.952838587109 42.12844209931791, 130.44879527622834 58.73342551570386, 131.89466654416174 75.44875323120505 M131.83588658459485 76.08132497780025 C102.30543430875986 76.48983121572063, 72.37445598617195 74.67762548385188, 24.056711124256253 73.50267897360027 M130.39905177708715 74.6283317739144 C104.6743649315089 76.90054796246812, 79.6103140341118 75.67280597595499, 24.334270439110696 75.56836631242186 M25.142140792682767 76.27820358984172 C25.568345656152815 62.0871600760147, 26.67828564858064 46.949199981987476, 26.775122391059995 23.932938857004046 M25.678110322915018 74.0633974308148 C23.806448485236615 59.39918356761336, 25.04840164957568 44.991463272832334, 25.94233041163534 24.231300049461424" stroke="#000" stroke-width="1" fill="none"/></g><text fill="rgb(27, 14, 139)" text-anchor="middle" font-family="Consolas, Courier" font-size="20" x="78" y="55">__main__</text><g><path d="M248.03387502096592 24.981137046590447 L447.86637596748767 25.23432659842074 L447.9792323675007 154.9256116244942 L248.39872595928608 154.70078069902956" stroke="none" stroke-width="0" fill="yellow"/><path d="M247.80511280181 24.848283371017935 C299.21462960343666 24.957142048644016, 350.4283732904477 25.27468179969294, 447.8333185797511 24.888723166230132 M248.04025084290154 25.084566500893157 C325.6484762636283 25.010388792416617, 403.7343389284854 25.26086524016505, 447.8703016929655 24.912154735801057 M447.72277866341176 24.848861547932028 C448.4073926393874 59.483614976145326, 447.8095053021796 94.24023997113109, 447.6624052260071 154.96266458891333 M448.1662653012201 25.192355196364225 C448.01118214936923 69.89683146812021, 447.9103734871931 114.68897119965405, 448.17893330883237 155.08975064624101 M448.1504552386168 155.19463287311416 C392.7703673179589 155.33646388878583, 337.31818679162086 155.01027628051926, 247.83021290746828 154.73049000131738 M447.8918324448064 154.93310165197937 C399.6892081104855 155.5109347502242, 351.60549481153026 155.28994757691015, 247.88017214083365 155.1023029807311 M248.02842815853654 155.25564071796833 C248.24258446125313 121.09422902893274, 248.46457245973872 86.7434340238571, 248.355024478212 24.78658777140081 M248.135622064583 154.81267948616295 C247.6169906966947 116.53043805956841, 247.86538132956252 78.29949534665792, 248.18846608232707 24.846260009892283" stroke="#000" stroke-width="4" fill="none"/></g><g><path d="M241.0472205001449 18.258276093946847 C296.7279231927182 18.81226224029811, 351.4515152491855 20.36467495903178, 453.1851150732587 18.45598127014297 M242.19678141163024 19.413435203408945 C323.6469059423271 19.100311479449005, 407.43214010978426 20.324859971231348, 453.3659209570659 18.57053532676623 M452.6138933170587 18.244307739660144 C456.0423227719031 56.304614026658236, 453.05288608586415 94.97427814751863, 452.3120261300355 160.8133229445666 M454.8313265061006 19.96177598182112 C454.07137157080695 67.92298510298133, 453.5673282599263 116.3225115230307, 454.89466654416174 161.44875323120505 M454.7355571239737 161.95153613589483 C395.8962872052512 162.68755860022557, 336.704565419905 161.09286757305927, 241.16993182411846 159.68239882282856 M453.4711821500076 160.6729422190877 C402.24132620213214 163.60327584438917, 351.59281165510396 162.52289678193645, 241.41417636055328 161.50014666802102 M242.14214079268277 162.27820358984172 C243.30960880378262 124.07270445320755, 244.41954879621045 84.92028873637318, 243.77512239106 17.932938857004046 M242.67811032291502 160.0633974308148 C239.97672923320906 118.29797793291509, 241.21868239754812 76.78905200343578, 242.94233041163534 18.231300049461424" stroke="rgb(0, 0, 0)" stroke-width="1" fill="none"/></g><text text-anchor="middle" font-family="Consolas, Courier" font-size="22px" fill="rgb(27, 14, 139)" font-weight="bolder" x="348" y="115">"David is cool!"</text><text fill="rgb(150, 100, 28)" text-anchor="middle" font-family="Consolas, Courier" font-size="22px" font-weight="bolder" x="278" y="55">id19</text><text fill="rgb(27, 14, 139)" text-anchor="middle" font-family="Consolas, Courier" font-size="22px" font-weight="bolder" x="418" y="55">str</text><g><path d="M247.7834524128586 24.83142109774053 C263.45769047215583 24.894601719416677, 278.9143965642899 25.247433857955038, 307.8147930715233 24.876355501636862 M248.0447244508192 25.09396549332887 C271.1546211728826 24.880231108143928, 294.79524127598853 25.15854631252587, 307.85588660668583 24.902391331084072 M307.72277866341176 24.848861547932028 C308.40024653943254 38.20590460542589, 307.80235920222475 51.68481922969222, 307.6624052260071 74.96266458891333 M308.1662653012201 25.192355196364225 C307.99056771742175 42.24940276928246, 307.88975905524563 59.39411380197853, 308.17893330883237 75.08975064624101 M308.167177316919 75.21626499556005 C291.52425968647003 75.26082695763557, 274.8012368466705 74.89838581126183, 247.81134222485124 74.70053579472005 M307.8798103554174 74.92566635478288 C293.39040774870665 75.28829942256212, 279.0331323316321 75.04275102525949, 247.86685408782213 75.11367326248437 M248.02842815853654 75.25564071796835 C248.11366913123055 61.97636435311288, 248.33565712971614 48.50770467221737, 248.355024478212 24.78658777140081 M248.135622064583 74.81267948616296 C247.76128969704732 60.09670382887125, 248.00968032991514 45.432026885263625, 248.18846608232707 24.846260009892283" stroke="#000" stroke-width="4" fill="none"/></g><g><path d="M387.7834524128586 24.83142109774053 C403.45769047215583 24.894601719416677, 418.9143965642899 25.247433857955038, 447.8147930715233 24.876355501636862 M388.0447244508192 25.09396549332887 C411.1546211728826 24.880231108143928, 434.79524127598853 25.15854631252587, 447.85588660668583 24.902391331084072 M447.72277866341176 24.848861547932028 C448.40024653943254 38.20590460542589, 447.80235920222475 51.68481922969222, 447.6624052260071 74.96266458891333 M448.1662653012201 25.192355196364225 C447.99056771742175 42.24940276928246, 447.88975905524563 59.39411380197853, 448.17893330883237 75.08975064624101 M448.167177316919 75.21626499556005 C431.52425968647003 75.26082695763557, 414.8012368466705 74.89838581126183, 387.8113422248513 74.70053579472005 M447.8798103554174 74.92566635478288 C433.39040774870665 75.28829942256212, 419.0331323316321 75.04275102525949, 387.86685408782216 75.11367326248437 M388.02842815853654 75.25564071796835 C388.11366913123055 61.97636435311288, 388.33565712971614 48.50770467221737, 388.355024478212 24.78658777140081 M388.135622064583 74.81267948616296 C387.7612896970473 60.09670382887125, 388.0096803299151 45.432026885263625, 388.18846608232707 24.846260009892283" stroke="#000" stroke-width="4" fill="none"/></g><g><path d="M484.0255640090501 24.241416855089682 C536.6747288307105 24.78571024322009, 588.3450280792927 26.373408998464704, 684.1665928987555 24.44361583115067 M485.20125421450774 25.42283250446578 C561.958478717734 25.051943962083083, 641.1038894416118 26.304326200825255, 684.3515084648274 24.560773679005287 M683.6138933170587 24.244307739660144 C687.036963196937 59.11295747105032, 684.047526510898 94.5909650363028, 683.3120261300355 154.8133229445666 M685.8313265061006 25.96177598182112 C685.0559107468464 69.77587079815567, 684.5518674359657 114.02828291337937, 685.8946665441617 155.44875323120505 M685.7522761930842 155.9731643655708 C630.2297164925149 156.68231944392923, 574.3466937635452 155.0513814025964, 484.15106453734137 153.65245000658692 M684.4591622240322 154.66550825989682 C636.1154268945954 157.554673751121, 588.366246741987 156.4497378845508, 484.4008607041682 155.51151490365555 M485.14214079268277 156.27820358984172 C486.2129223062657 121.20502475183457, 487.3228622986935 85.18492933362721, 486.77512239106 23.932938857004046 M485.678110322915 154.0633974308148 C483.0849534834735 115.83291779831052, 484.32690664781256 77.85893173422663, 485.94233041163534 24.231300049461424" stroke="#000" stroke-width="1" fill="none"/></g><g><path d="M478.0472205001449 18.258276093946847 C533.7279231927182 18.81226224029811, 588.4515152491855 20.36467495903178, 690.1851150732587 18.45598127014297 M479.19678141163024 19.413435203408945 C560.6469059423271 19.100311479449005, 644.4321401097843 20.324859971231348, 690.3659209570659 18.57053532676623 M689.6138933170587 18.244307739660144 C693.0423227719032 56.304614026658236, 690.0528860858642 94.97427814751863, 689.3120261300355 160.8133229445666 M691.8313265061006 19.96177598182112 C691.071371570807 67.92298510298133, 690.5673282599263 116.3225115230307, 691.8946665441617 161.44875323120505 M691.7355571239738 161.95153613589483 C632.8962872052513 162.68755860022557, 573.7045654199051 161.09286757305927, 478.16993182411846 159.68239882282856 M690.4711821500076 160.6729422190877 C639.2413262021321 163.60327584438917, 588.5928116551039 162.52289678193645, 478.4141763605533 161.50014666802102 M479.14214079268277 162.27820358984172 C480.3096088037826 124.07270445320755, 481.41954879621045 84.92028873637318, 480.77512239106 17.932938857004046 M479.678110322915 160.0633974308148 C476.976729233209 118.29797793291509, 478.2186823975481 76.78905200343578, 479.94233041163534 18.231300049461424" stroke="rgb(0, 0, 0)" stroke-width="1" fill="none"/></g><text text-anchor="middle" font-family="Consolas, Courier" font-size="20" fill="rgb(27, 14, 139)" x="585" y="115">7</text><text fill="rgb(150, 100, 28)" text-anchor="middle" font-family="Consolas, Courier" font-size="20" x="515" y="55">id13</text><text fill="rgb(27, 14, 139)" text-anchor="middle" font-family="Consolas, Courier" font-size="20" x="655" y="55">int</text><g><path d="M483.917262064293 24.157105488702655 C500.9689266048372 24.47300859708339, 516.9329313095659 26.237169289775192, 544.0739653576165 24.381777508184314 M485.223622254096 25.469827466644347 C507.3879350842908 24.40115554071963, 532.205864819698 25.792731562629342, 544.2794330334291 24.511956655420363 M543.6138933170587 24.244307739660144 C547.0012326971628 37.83524710033089, 544.0117960111238 52.03554429486394, 543.3120261300355 74.81332294456661 M545.8313265061006 25.96177598182112 C544.952838587109 42.12844209931791, 544.4487952762283 58.73342551570386, 545.8946665441617 75.44875323120505 M545.8358865845948 76.08132497780025 C529.0346624031663 76.3041347881779, 511.8329121749848 74.49192905630916, 484.05671112425625 73.50267897360027 M544.3990517770872 74.6283317739144 C529.7528546461836 76.44149711281061, 515.7672934634611 75.21375512629747, 484.3342704391107 75.56836631242186 M485.14214079268277 76.27820358984172 C485.5683456561528 62.0871600760147, 486.67828564858064 46.949199981987476, 486.77512239106 23.932938857004046 M485.678110322915 74.0633974308148 C483.8064484852366 59.39918356761336, 485.0484016495757 44.991463272832334, 485.94233041163534 24.231300049461424" stroke="#000" stroke-width="1" fill="none"/></g><g><path d="M623.917262064293 24.157105488702655 C640.9689266048372 24.47300859708339, 656.9329313095659 26.237169289775192, 684.0739653576165 24.381777508184314 M625.223622254096 25.469827466644347 C647.3879350842908 24.40115554071963, 672.205864819698 25.792731562629342, 684.2794330334291 24.511956655420363 M683.6138933170587 24.244307739660144 C687.0012326971628 37.83524710033089, 684.0117960111238 52.03554429486394, 683.3120261300355 74.81332294456661 M685.8313265061006 25.96177598182112 C684.952838587109 42.12844209931791, 684.4487952762283 58.73342551570386, 685.8946665441617 75.44875323120505 M685.8358865845948 76.08132497780025 C669.0346624031663 76.3041347881779, 651.8329121749848 74.49192905630916, 624.0567111242563 73.50267897360027 M684.3990517770872 74.6283317739144 C669.7528546461836 76.44149711281061, 655.7672934634611 75.21375512629747, 624.3342704391107 75.56836631242186 M625.1421407926828 76.27820358984172 C625.5683456561528 62.0871600760147, 626.6782856485806 46.949199981987476, 626.77512239106 23.932938857004046 M625.678110322915 74.0633974308148 C623.8064484852366 59.39918356761336, 625.0484016495757 44.991463272832334, 625.9423304116353 24.231300049461424" stroke="#000" stroke-width="1" fill="none"/></g></svg>"`;
